examples/hello_compute_c/main.c

```diff
@@ -0,0 +1,139 @@
+        &(WGPUBufferDescriptor){.size = numbers_length,
+            .usage = WGPUBufferUsage_MAP_READ | WGPUBufferUsage_TRANSFER_DST |
+                WGPUBufferUsage_TRANSFER_SRC},
+            &staging_memory);
+
+    WGPUBufferId storage_buffer = wgpu_device_create_buffer(device,
+        &(WGPUBufferDescriptor){.size = size,
+            .usage = WGPUBufferUsage_STORAGE | WGPUBufferUsage_TRANSFER_DST |
+                WGPUBufferUsage_TRANSFER_SRC});
+
+    WGPUBindGroupLayoutId bind_group_layout =
+        wgpu_device_create_bind_group_layout(device,
+            &(WGPUBindGroupLayoutDescriptor){
+                .bindings = &(WGPUBindGroupLayoutBinding){.binding = 0,
+                    .visibility = WGPUShaderStage_COMPUTE,
+                    .ty = WGPUBindingType_StorageBuffer}});
+
+	WGPUBindingResource resource = (WGPUBindingResource) {
+        .tag = WGPUBindingResource_Buffer,
+        .buffer = (WGPUBindingResource_WGPUBuffer_Body) {
+            storage_buffer
+        }};
+
+    WGPUBindGroupId bind_group = wgpu_device_create_bind_group(device,
+            &(WGPUBindGroupDescriptor){.layout = bind_group_layout,
+                .bindings = &(WGPUBindGroupBinding){
+                    .binding = 0, .resource = resource}});
+
+    WGPUPipelineLayoutId pipeline_layout =
+        wgpu_device_create_pipeline_layout(device,
+            &(WGPUPipelineLayoutDescriptor){
+                .bind_group_layouts = &bind_group_layout});
+
+    WGPUShaderModuleId shader_module = wgpu_device_create_shader_module(device,
+        &(WGPUShaderModuleDescriptor){
+            .code = read_file("./../../data/shader.comp.spv")});
+
+    WGPUComputePipelineId compute_pipeline =
+        wgpu_device_create_compute_pipeline(device,
+            &(WGPUComputePipelineDescriptor){.layout = pipeline_layout,
+                .compute_stage = (WGPUPipelineStageDescriptor){
+                    .module = shader_module, 
+					.entry_point = "main"}});
+
+    WGPUCommandEncoderId encoder = wgpu_device_create_command_encoder(
+        device, &(WGPUCommandEncoderDescriptor){.todo = 0});
+
+    wgpu_command_buffer_copy_buffer_to_buffer(
+        encoder, staging_buffer, 0, storage_buffer, 0, size);
+
+    WGPUComputePassId command_pass =
+        wgpu_command_encoder_begin_compute_pass(encoder);
+    wgpu_compute_pass_set_pipeline(command_pass, compute_pipeline);
+
+    wgpu_compute_pass_set_bind_group(command_pass, 0, bind_group, 0, 0);
+    wgpu_compute_pass_dispatch(command_pass, numbers_length, 1, 1);
+
+    wgpu_command_buffer_copy_buffer_to_buffer(
+        encoder, storage_buffer, 0, staging_buffer, 0, size);
+
+    WGPUQueueId queue = wgpu_device_get_queue(device);
+
+    WGPUCommandBufferId command_buffer = wgpu_command_encoder_finish(encoder);
+
+    wgpu_queue_submit(queue, &command_buffer, 1);
+
+    wgpu_buffer_map_read_async(staging_buffer, 0, size, &read_buffer_map, NULL);
+
```

Comment:
In order for `read_buffer_map` to actually execute, we have to wait for the GPU to run the compute shader to update storage buffer, then copy storage buffer to staging buffer, then call our callback. So we need a way to indicate to the GPU that we want to wait until it has finished.

We can do this with `wgpu_device_poll(device, true);`. This will give the chance wgpu to check pending commands and perform any necessary work. The second argument `true` indicates that we want to wait until all work is done on the GPU, at which point our `read_buffer_map` callback will be executed.